### PR TITLE
refactor to move more processing to the API

### PR DESCRIPTION
**Pull request by grokspawn**
**Description of the change:**
move all parsing and template logic into the API, out of the cobra command parser
facilitate testing of the HTTP access for catalog files

to accomplish this, we exposed a dynamic set of builders as a property of the template, but we didn't want to export it, so that meant unexporting all template fields and leveraging the options pattern to provide access to external callers.  This meant that we could mock out the builder interactions with dummy objects to limit test interactions to just this package.  
In addition, we moved the http/filesystem processing for the catalogs.yaml access to an exported function so we could utest interactions here.  

**Motivation for the change:**
to facilitate ability to test with different input file/URI

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: pkg/cache/pkgs.go
Comment (everettraven) [inline pkg/cache/pkgs.go:1]: Out of scope?

**Review by everettraven — CHANGES_REQUESTED** [review]
Overall looks good. I have a concern about using `cmd.CombinedOutput` vs `cmd.Output` and a few other nits/comments

File: alpha/template/composite/builder.go
Comment (grokspawn) [reply alpha/template/composite/builder.go:267]: Fair.  The origin of this change was when this version of opm would shell out to a specified containerized opm version in a manner that caused that binary to generate a usage statement and quit with a zero exit status.  That meant that we discarded all remediation data by the time we tried to parse the data (and failed), so capturing the STDERR was the only way to be sure. 

I think it's likely more beneficial to service the more likely use-case of testing for error status rather than fix a (now obsolete) pretty obscure edge case.

File: cmd/opm/alpha/template/composite.go
Comment (oceanc80) [reply cmd/opm/alpha/template/composite.go:36]: +1

Comment (grokspawn) [reply cmd/opm/alpha/template/composite.go:36]: I'll push back on this one, because I noticed it too but made the conscious choice to perpetuate a common pattern which was also demonstrated in
cmd/opm/init/cmd.go
cmd/opm/render/cmd.go
cmd/opm/alpha/template/basic.go

Comment (oceanc80) [reply cmd/opm/alpha/template/composite.go:36]: Right, but in all three of those cases there is custom logic for the yaml and json cases vs a no-op.
e.g. https://github.com/operator-framework/operator-registry/blob/master/cmd/opm/alpha/template/basic.go#L41-L48

Comment (everettraven) [reply cmd/opm/alpha/template/composite.go:36]: +1 to @oceanc80's point

Comment (everettraven) [reply cmd/opm/alpha/template/composite.go:36]: Not a huge deal though, so if you prefer leaving this as is I'm okay with it

File: alpha/template/composite/composite.go
Comment (everettraven) [inline alpha/template/composite/composite.go:218]: Could be done in a follow up, but is the `.Destination.BaseImage` field still necessary since we aren't running `opm` images anymore? If the image isn't going to be taken into account by the rendering process we should probably remove it (or at least comment it out if we plan to use it later) to prevent confusion

**Review by everettraven — APPROVED** [review]
Looks good to me. Have a comment that could be addressed now or in a follow up - I'll leave it up to you.

Going to hold off on adding the LGTM label to prevent an auto-merge so others have a chance to review

File: alpha/template/composite/composite.go
Comment (grokspawn) [reply alpha/template/composite/composite.go:218]: I agree that this should be commented-out, since I don't think discussion is dead on it right now, but we definitely don't want to enforce it.  
However, because dependencies are starting to stack up against this PR refactor, I'd prefer to handle in another PR.

**Pull request merged**
